klippy/extras/z_thermal_adjust.py

```diff
@@ -0,0 +1,195 @@
+        self.gcode.register_command('SET_Z_THERMAL_ADJUST_REF',
+                                    self.cmd_SET_Z_THERMAL_ADJUST_REF,
+                                    desc=self.cmd_SET_Z_THERMAL_ADJUST_REF_help)
+
+    def handle_connect(self):
+        'Called after all printer objects are instantiated'
+        self.toolhead = self.printer.lookup_object('toolhead')
+        gcode_move = self.printer.lookup_object('gcode_move')
+
+        # Temperature sensor config check and callback registration
+        try:
+            self.sensor = self.printer.lookup_object(self.temp_sensor_name)
+        except Exception as e:
+            msg = '''
+            %s\nUse the full config section name in the "temp_sensor" parameter,
+            e.g. "temp_sensor: temperature_sensor frame".
+            ''' % e
+            raise self.printer.config_error(msg)
+        else:
+            self.sensor.sensor.setup_callback(self.z_adj_temperature_callback)
+
+        # Register move transformation
+        self.next_transform = gcode_move.set_move_transform(self, force=True)
+
+        # Pull Z step distance for minimum adjustment increment
+        kin = self.printer.lookup_object('toolhead').get_kinematics()
+        steppers = [s.get_name() for s in kin.get_steppers()]
+        z_stepper = kin.get_steppers()[steppers.index("stepper_z")]
+        self.z_step_dist = z_stepper.get_step_dist()
+
+    def get_status(self, eventtime):
+        return {
+            'temperature': self.smoothed_temp,
+            'current_z_adjust': self.z_adjust_mm,
+            'z_adjust_ref_temperature': self.ref_temperature,
+            'enabled': self.adjust_enable
+        }
+
+    def handle_homing_move_end(self, homing_state, rails):
+        'Set reference temperature after Z homing.'
+        if 2 in homing_state.get_axes():
+            self.ref_temperature = self.smoothed_temp
+            self.z_adjust_mm = 0.
+
+    def calc_adjust(self, pos):
+        'Z adjustment calculation'
+        if pos[2] < self.off_above_z:
+            delta_t = self.smoothed_temp - self.ref_temperature
+
+            # Calculate Z adjustment
+            adjust = -1 * self.temp_coeff * delta_t
+
+            # compute sign (+1 or -1) for maximum offset setting
+            sign = 1 - (adjust <= 0)*2
+
+            # Don't apply adjustments smaller than step distance
+            if abs(adjust - self.z_adjust_mm) > self.z_step_dist:
+                self.z_adjust_mm = min([self.max_z_adjust_mm*sign,
+                    adjust], key=abs)
+
+        # Apply Z adjustment
+        new_z = pos[2] + self.z_adjust_mm
+        self.last_z_adjust_mm = self.z_adjust_mm
+        return [pos[0], pos[1], new_z, pos[3]]
+
+    def calc_unadjust(self, pos):
+        'Remove Z adjustment'
+        unadjusted_z = pos[2] - self.z_adjust_mm
+        return [pos[0], pos[1], unadjusted_z, pos[3]]
+
+    def get_position(self):
+        return self.calc_unadjust(self.next_transform.get_position())
```

Comment:
I'm confused why this doesn't set `self.last_position`?

Reply:
Forgive my ignorance, is that just an expected property for a module like this?

Reply:
Forgive my ignorance, is that just an expected property for a module like this?

Reply:
A gcode transform must be able to both translate to a new coordinate system (`move()`) and report the current position in the old coordinate system (`get_position()`).  If some external process changes the position outside of a normal gcode `move()` call, then the system arranges for `get_postiion()` to be called so that all components in the gcode stream know the current position.

It seems the code is tracking a `last_position`, but it isn't resetting it on a `get_position()` call.  That seems suspect to me.

-Kevin

Reply:
Thank you for the explanation, Kevin!